File: epic_ops/epic_ops/iou.py
import torch


@torch.no_grad()
def instance_seg_iou_csr(
    proposal_indices: torch.Tensor,
    instance_labels: torch.Tensor,
    num_points_per_instance: torch.Tensor,
) -> torch.Tensor:
    proposal_indices = proposal_indices.contiguous()
    instance_labels = instance_labels.contiguous()
    num_points_per_instance = num_points_per_instance.contiguous()

    return torch.ops.epic_ops.instance_seg_iou_csr(
        proposal_indices, instance_labels, num_points_per_instance
    )


@torch.no_grad()
def instance_seg_iou(
    proposal_indices_begin: torch.Tensor,
    proposal_indices_end: torch.Tensor,
    instance_labels: torch.Tensor,
    num_points_per_instance: torch.Tensor,
) -> torch.Tensor:
    proposal_indices_begin = proposal_indices_begin.contiguous()
    proposal_indices_end = proposal_indices_end.contiguous()
    instance_labels = instance_labels.contiguous()
    num_points_per_instance = num_points_per_instance.contiguous()

    return torch.ops.epic_ops.instance_seg_iou(
        proposal_indices_begin, proposal_indices_end,
        instance_labels, num_points_per_instance,
    )


@torch.no_grad()
def batch_instance_seg_iou(
    proposal_offsets: torch.Tensor,
    instance_labels: torch.Tensor,
    batch_indices: torch.Tensor,
    num_points_per_instance: torch.Tensor,
) -> torch.Tensor:
    proposal_offsets = proposal_offsets.contiguous()
    instance_labels = instance_labels.contiguous()
    batch_indices = batch_indices.contiguous()
    num_points_per_instance = num_points_per_instance.contiguous()

    return torch.ops.epic_ops.batch_instance_seg_iou(
        proposal_offsets, instance_labels, batch_indices, num_points_per_instance
    )

File: epic_ops/src/epic_ops/iou.cpp
#include <torch/library.h>
#include "epic_ops/iou.h"

namespace epic_ops::iou {

at::Tensor instance_seg_iou_csr(
    const at::Tensor& proposal_indices,
    const at::Tensor& instance_labels,
    const at::Tensor& num_points_per_instance) {
  static auto op = c10::Dispatcher::singleton()
                       .findSchemaOrThrow("epic_ops::instance_seg_iou_csr", "")
                       .typed<decltype(instance_seg_iou_csr)>();
  return op.call(proposal_indices, instance_labels, num_points_per_instance);
}

at::Tensor instance_seg_iou(
    const at::Tensor& proposal_indices_begin,
    const at::Tensor& proposal_indices_end,
    const at::Tensor& instance_labels,
    const at::Tensor& num_points_per_instance) {
  static auto op = c10::Dispatcher::singleton()
                       .findSchemaOrThrow("epic_ops::instance_seg_iou", "")
                       .typed<decltype(instance_seg_iou)>();
  return op.call(proposal_indices_begin, proposal_indices_end, instance_labels, num_points_per_instance);
}

at::Tensor batch_instance_seg_iou(
    const at::Tensor& proposal_offsets,
    const at::Tensor& instance_labels,
    const at::Tensor& batch_indices,
    const at::Tensor& num_points_per_instance) {
  static auto op = c10::Dispatcher::singleton()
                       .findSchemaOrThrow("epic_ops::batch_instance_seg_iou", "")
                       .typed<decltype(batch_instance_seg_iou)>();
  return op.call(
      proposal_offsets, instance_labels, batch_indices, num_points_per_instance);
}

TORCH_LIBRARY_FRAGMENT(epic_ops, m) {
  m.def(TORCH_SELECTIVE_SCHEMA(
      "epic_ops::instance_seg_iou_csr(Tensor proposal_indices, Tensor instance_labels, "
      "Tensor num_points_per_instance) -> Tensor"));

  m.def(TORCH_SELECTIVE_SCHEMA(
      "epic_ops::instance_seg_iou(Tensor proposal_indices_begin, Tensor proposal_indices_end, "
      "Tensor instance_labels, Tensor num_points_per_instance) -> Tensor"));

  m.def(TORCH_SELECTIVE_SCHEMA(
      "epic_ops::batch_instance_seg_iou(Tensor proposal_offsets, Tensor instance_labels, "
      "Tensor batch_indices, Tensor num_points_per_instance) -> Tensor"));
}

} // namespace epic_ops::iou
